### ISSUE TITLE
Seeking advice for Python wrapper for cbSdkGetSpkCache

### ISSUE THREAD

**Issue by cboulay** [dashesy/CereLink#76]
I'm just about done modifying the Python wrapper to expose all the functionality I need for my GUI. I hope to make a PR for that in the near future. ([Example PR here](https://github.com/SachsLab/CereLink/pull/2); most of the changes are converting MS types to STL types (UINT8->uint8_t))

Edit:
I wrote a [Cython wrapper class for calling `cbSdkGetSpkCache`](https://github.com/SachsLab/CereLink/blob/Production/cerebus/cbpy.pyx#L782-L817). The model requires the user to maintain a SpikeCache instance for each channel. i.e.

```Python
# somewhere
self.spike_cache = {}

# Later
for chan_ix in range(n_channels):
    if chan_ix not in self.spike_cache:
        self.spike_cache[chan_ix] = cbpy.SpikeCache(channel=chan_ix, instance=0)

    # Then, when fetching data
    wfs, unit_ids = self.spike_cache[chan_ix].get_new_waveforms()
```

I would be interested to hear any other ideas on how the spk cache should be exposed to the Python user.

Note: Central cannot start if an application owning a `cbSPKCACHE` is already running on the same computer.

**Comment by cboulay** [comment]
I was able to confirm that even a simple C++ application owning a pointer to `cbSPKCACHE`, using the official Blackrock SDK, prevents Central from starting. However, if Central is already running then this doesn't seem to be a problem. It's worth noting that my application connects as `CBSDKCONNECTION_UDP` if Central is not running or as `UDPCBSDKCONNECTION_CENTRAL` if it is running.

The test code is below.

```C++
#include "stdafx.h"
#define NO_AFX  // Because my Windows 10 + MSVC C++ 2010 Express combination is missing some types.
#include "cbsdk.h"
#include <string>
#include <iostream>

int _tmain(int argc, _TCHAR* argv[])
{
	cbSdkResult res;
	cbSdkConnection con = cbSdkConnection();
	cbSdkConnectionType conType = CBSDKCONNECTION_DEFAULT; //CBSDKCONNECTION_DEFAULT, UDPCBSDKCONNECTION_CENTRAL, CBSDKCONNECTION_UDP
	res = cbSdkOpen(0, conType, con);
	cbSdkVersion ver;
	res = cbSdkGetVersion(0, &ver);
	cbSdkInstrumentType instType;
	std::string devString;
	if (res == CBSDKRESULT_SUCCESS)
	{
		res = cbSdkGetType(0, &conType, &instType);
		if (res == CBSDKRESULT_SUCCESS)
		{
			if (instType != CBSDKINSTRUMENT_LOCALNSP && instType != CBSDKINSTRUMENT_NSP)
				devString = "nPlay";
			if (conType == CBSDKCONNECTION_CENTRAL)
				devString += "(Central)";
		}
		std::cout << devString << " v" << ver.nspmajor << "." << ver.nspminor << "." << ver.nsprelease << "." << ver.nspbeta << std::endl;

		cbSPKCACHE ignoreme;  // cbSdkGetSpkCache modifies what pt_cache is pointing to, so this variable is here only to avoid 'new' with missing 'delete'.
		cbSPKCACHE *pt_cache = &ignoreme;
		res = cbSdkGetSpkCache(0, 1, &pt_cache);
		UINT32 start_valid = pt_cache->valid;
		while (pt_cache->valid < (start_valid + 300))
		{
			std::cout << "chid:" << pt_cache->chid << "\thead:" << pt_cache->head << "\tpktcnt:" << pt_cache->pktcnt << "\tpktsize:" << pt_cache->pktsize << "\tvalid:" << pt_cache->valid << std::endl;
			Sleep(100);
		}
	}
    return 0;
}
```

**Comment by cboulay** [comment]
I made quite a few changes to how I'm using `cbSdkGetSpkCache`. I modified the original post to reflect the new state. I'll make my PR at the end of the week if there are no objections here.

**Comment by dashesy** [comment]
Central is not well tested as client, it assumes ownership. I made sure it runs (as client) for normal cbsdk usage but never tested it when spike cache is owned by cbsdk.

**Comment by cboulay** [comment]
For now it is predictable and mostly stable. One way I can reliably get my spike cache visualization program to crash is if I use another program to modify the spike threshold.

**Comment by dashesy** [comment]
The problem is with access to the raw pointer, while it is being modified. The proper access should double buffer and use a lock. I will have to look at how the cache is populated to know why it may crash when changing spike threshold. It should not crash

**Comment by cboulay** [comment]
The SpkCache crash was fixed in 7fba59ca24e56fb4854318c711deae448833dd37
